File: java_version/mixed_recall/middle_value/middle_value.java
package mixed_recall.middle_value;
// middle value

// Write a function, middleValue, 
// that takes in the head of a linked list as an argument. 
// The function should return the value of the middle node in the linked list. 
// If the linked list has an even number of nodes, then return the value of the second middle node.

// You may assume that the input list is non-empty.

import java.util.ArrayList;
import java.util.List;

class Node<T> {
    T val;
    Node<T> next;

    public Node(T val) {
        this.val = val;
        this.next = null;
    }
}

public class middle_value {
    public static <T> String middleValueArray(Node<T> head) {
    
        return "";
    }

    public static <T> String middleValueDualPointers(Node<T> head) {
        return "";
        
    }

    public static void main(String[] args) {
        Node<String> a = new Node<>("a");
        Node<String> b = new Node<>("b");
        Node<String> c = new Node<>("c");
        Node<String> d = new Node<>("d");
        Node<String> e = new Node<>("e");

        a.next = b;
        b.next = c;
        c.next = d;
        d.next = e;

        // a -> b -> c -> d -> e
        System.out.println(middleValueArray(a)); // c
        System.out.println(middleValueDualPointers(a)); // c

        // Node a("a");
        // Node b("b");
        // Node c("c");
        // Node d("d");
        // Node e("e");
        Node<String> f = new Node<>("f");

        a.next = b;
        b.next = c;
        c.next = d;
        d.next = e;
        e.next = f;

        // a -> b -> c -> d -> e -> f
        System.out.println(middleValueArray(a)); // d
        System.out.println(middleValueDualPointers(a)); // d

        Node<String> x = new Node<>("x");
        Node<String> y = new Node<>("y");
        Node<String> z = new Node<>("z");

        x.next = y;
        y.next = z;

        // x -> y -> z
        System.out.println(middleValueArray(x)); // y
        System.out.println(middleValueDualPointers(x)); // y

        // Node x("x");
        // Node y("y");

        x.next = y;
        y.next = null;
        // x -> y
        System.out.println(middleValueArray(x)); // y
        System.out.println(middleValueDualPointers(x)); // y

        Node<String> q = new Node<>("q");

        // q
        System.out.println(middleValueArray(q)); // q
        System.out.println(middleValueDualPointers(q)); // q

    }
}

File: javascript_version/mixed_recall/middle_value/middle_value.js
// middle value
// Write a function, middleValue, 
// that takes in the head of a linked list as an argument. 
// The function should return the value of the middle node in the linked list. 
// If the linked list has an even number of nodes, then return the value of the second middle node.

// You may assume that the input list is non-empty.


class Node {
    constructor (val) {
        this.val = val;
        this.next = null;
    }
}

/*
n = number of nodes
Time: O(n)
Space: O(n)*/
const middleValueArray = (head) => {
    let result = [];
    let curr = head;
    while (curr !== null) {
        result.push(curr.val);
        curr = curr.next;
    }
    return result[Math.floor(result.length / 2)];
};

/*n = number of nodes
Time: O(n)
Space: O(1)*/
const middleValueDualPointers = (head) => {
    let slow = head;
    let fast = head;
    while (fast !== null && fast.next !== null) {
        slow = slow.next;
        fast = fast.next.next;
    }
    return slow.val;
};


const a = new Node('a');
const b = new Node('b');
const c = new Node('c');
const d = new Node('d');
const e = new Node('e');

a.next = b;
b.next = c;
c.next = d;
d.next = e;

// a -> b -> c -> d -> e
console.log(middleValueArray(a)); // c
console.log(middleValueDualPointers(a)); // c

// const a = new Node('a');
// const b = new Node('b');
// const c = new Node('c');
// const d = new Node('d');
// const e = new Node('e');
const f = new Node('f');

a.next = b;
b.next = c;
c.next = d;
d.next = e;
e.next = f;

// a -> b -> c -> d -> e -> f
console.log(middleValueArray(a)); // d
console.log(middleValueDualPointers(a)); // d

const x = new Node('x');
const y = new Node('y');
const z = new Node('z');

x.next = y;
y.next = z;

// x -> y -> z
console.log(middleValueArray(x)); // y
console.log(middleValueDualPointers(x)); // y

// const x = new Node('x');
// const y = new Node('y');

x.next = y;
y.next = null;
// x -> y 
console.log(middleValueArray(x)); // y
console.log(middleValueDualPointers(x)); // y

const q = new Node('q');

// q
console.log(middleValueArray(q)); // q
console.log(middleValueDualPointers(q)); // q


